compiler/rustc_lint/src/levels.rs

```diff
@@ -116,8 +116,29 @@ impl LintLevelSets {
     }
 }
 
+fn lints_that_will_run_for_id(tcx: TyCtxt<'_>, key: OwnerId) -> Vec<LintId> {
+    let shallow_lint_level = tcx.shallow_lint_levels_on(key);
+
+    let mut needs_to_run = Vec::new();
+
+    for (_, specs) in shallow_lint_level.specs.iter() {
+        for (lint, level_and_source) in specs.iter() {
+            if !matches!(level_and_source.level, Level::Allow) {
+                needs_to_run.push(*lint)
+            }
+        }
+    }
+
+    needs_to_run
+}
+
 fn lints_that_dont_need_to_run(tcx: TyCtxt<'_>, (): ()) -> UnordSet<LintId> {
```

Comment:
A few ideas:
- make the return type a dedicated struct;
- replace `dont_need_to_run.contains(lint)` by a call to inherent method `should_run(lint)`;
- perform the check for future breakage and `lint_cap` in that method;
- compute this query per module, and only recurse inside that module.